File: linked_list.rb
=begin
linked_list.rb
Anh Uong 
Fall 2015

creates a linked list data structure composed of nodes
  - uses classes as basis with functions inside of class
  - singly-linked list with head, tail, and num_nodes
=end

class Node 
  
  def initialize(value)
    @data = value
    @next = nil
  end
  
  def get_Data()
    return @data
  end
  
  def set_Data(value)
    @data = value
  end
  
  def get_Next()
    return @next
  end
  
  def set_Next(next_node)
    @next = next_node
    return @next
  end
  
end

class Linked_List
  
  def initialize()
    @head = nil
    @tail = nil
    @num_nodes = 0
  end
  
  # adds node to FRONT of list, storing given data in the node.
  def ll_push(data)
    new_node = Node.new(data)
    if @num_nodes == 0
      @head = new_node
      @tail = new_node
    else
      current = @head
      new_node.set_Next(current)
      @head = new_node
    end
    
    @num_nodes += 1
  end
  
  # removes node at FRONT of list and returns corresponding data
  def ll_pop()
    current = @head
    
    if @num_nodes == 0
      return nil
    elsif @num_nodes == 1
      @head = nil
      @tail = nil
    else
      second = @head.get_Next()
      @head = second
    end
    
    @num_nodes -= 1
    return current.get_Data()
  end
   
   # adds node to end of list, storing given data in node
  def ll_append(data)
     new_node = Node.new(data)
     
     if @num_nodes == 0
       @head = new_node
       @tail = new_node
     else
       end_node = @tail
        end_node.set_Next(new_node)
       @tail = new_node
     end
    
     @num_nodes += 1
  end
   
   # removes first node in list THAT MATCHES target
   # returns pointer to the data UNLESS node does not exist (returns nil)
  def ll_remove(target)
    target_node = find_node(target)

    if target_node == nil
     return nil
    end

    # if first node is being removed -> need to reset head to node 
    if target_node[0] == 0
     if @num_nodes == 1
       @head = nil
       @tail = nil
     else
       @head = @head.get_Next()
     end
    # if any other node in middle
    else
     current = @head
     # find node that comes before one being deleted
     for i in (0..(target_node[0]-2))
       current = current.get_Next()
     end
     
     # if last node -> need to reset tail to node prior
     if target_node[0] == @num_nodes - 1
       current.set_Next(nil)
       @tail = current
     else
       current.set_Next(current.get_Next().get_Next())
     end
    end

     @num_nodes -= 1
     return target_node[1].get_Data();
    end

    # traverses linked list searching for node that matches target
    # returns pointer to node w/ value unless doesn't exist (returns nil)
  def find_node(target)
   if @num_nodes == 0
     return nil
   else
     current = 0
     current_node = @head
     while current < @num_nodes
       if current_node.get_Data() == target
         return current, current_node
       end
 
       current_node = current_node.get_Next()
       current += 1
     end
     return nil
   end
  end

  # removes all nodes from list
  def ll_clear()
   @head = nil
   @tail = nil
   @num_nodes = 0
  end

  # traverses list and applies given function to the data in each node
  def ll_map(function) 
   current = 0
   current_node = @head

   while current < @num_nodes
     current_node.set_Data(function.call(current_node.get_Data()))
     current_node = current_node.get_Next()
     current += 1
   end
  end
  
  # returns size of linked list
  def ll_size()
   return @num_nodes
  end
  
  # returns NODE head is pointing to
  def get_Head()
   return @head
  end
  
  # returns NODE tail is pointing to
  def get_Tail()
   return @tail
  end
 
end

# creates new LinkedList, initializes, and returns
def ll_create()
  return Linked_List.new()
end

 
# initialize new Linked List 
ll = Linked_List.new()
# could also do ll = ll_create()

# add differing data type values to Linked list
ll.ll_push(50)
ll.ll_push("hi")
ll.ll_append(false)
ll.ll_append(638.3)

# print out values of nodes in order
# LinkedList = "hi", 50, false, 638.3
print "A: LinkedList has ", ll.ll_size(), " nodes\n"
current = ll.get_Head()
for i in (1..ll.ll_size())
  print "Index ", i, " in LinkedList = ", current.get_Data(), "\n"
  current = current.get_Next()
end

# print and delete certain values/nodes
print "\nRemoving 10 -> ", ll.ll_remove(10), "\n"     # nil (aka blank)
print "Removing false -> ", ll.ll_remove(false), "\n"  # false (which means able to delete)
print "Popping node -> ", ll.ll_pop(), "\n"          # "hi" 

# print out values of nodes in order
# LinkedList = 50, 638.3
print "\nB: LinkedList has ", ll.ll_size(), " nodes\n"
current = ll.get_Head()
for i in (1..ll.ll_size())
  print "Index ", i, " in LinkedList = ", current.get_Data(), "\n"
  current = current.get_Next()
end

# empties list
ll.ll_clear()
print "\nC: Clearing list - LinkedList has ", ll.ll_size(), " nodes\n"

# create two linked lists with different data types as values
l2 = Linked_List.new()
ll.ll_push(1)
ll.ll_push(2)
ll.ll_push(3)
ll.ll_push(4)
ll.ll_push(5)
# ll = 5, 4, 3, 2, 1

l2.ll_append(1.1)
l2.ll_append(2.2)
l2.ll_append(3.3)
l2.ll_append(4.4)
l2.ll_append(5.5)
# l2 = 1.1, 2.2, 3.3, 4.4, 5.5

# squares value within node
def square(value)
  puts value**2
  return value**2
end

def print(value)
  puts value
  return value
end

# applies above function to all nodes via mapping function
puts "\nLinked List 1 values:"
ll.ll_map(method(:print))
puts "Linked List 2 values:"
l2.ll_map(method(:print))

puts "\nLinked List 1 values squared:"
ll.ll_map(method(:square))
puts "Linked List 2 values squared:"
l2.ll_map(method(:square))

# ensures that values were changed - prints out values in nodes
puts "\nEnsure worked - Linked List 1 values:"
ll.ll_map(method(:print))

    
      
    

File: linked_list.php
<?php
/*
linked_list.php
Anh Uong 
November 2, 2015

creates a linked list data structure composed of nodes
  - uses classes as basis with functions inside of class
  - singly-linked list with head, tail, and num_nodes
=end
*/


class Node{
	private $data;
	private $next;
	
	function __construct($value=NULL, $next_node=NULL){
		$this->data = $value;
		$this->next = $next_node;
	}
	
	function get_Data(){
		return $this->data;
	}
	
	function set_Data($value){
		$this->data = $value;
	}
	
	function get_Next(){
		return $this->next;
	}
	
	function set_Next($next_node){
		$this->next = $next_node;
	}
}


class Linked_List{
	
	private $head;
	private $tail;
	private $num_nodes;
	
	function __construct(){
		$this->head = NULL;
		$this->tail = NULL;
		$this->num_nodes = 0;
	}
	
    # adds node to FRONT of list, storing given data in the node.
    function ll_push($data){
    	$node = new Node($data);
		if($this->num_nodes == 0){
			$this->head = $node;
			$this->tail = $node;
		}
		else{
			$current = $this->head;
			$node->set_Next($current);
			$this->head = $node;
				
		}
		$this->num_nodes++;
    }
	
    # removes node at FRONT of list and returns corresponding data
    function ll_pop(){
    	$current = $this->head;
		
		if( $this->num_nodes == 0){
			return NULL;
		}
		elseif($this->num_nodes == 1){
			$this->head = NULL;
			$this->tail = NULL;
		}
		else{
			$this->head = $this->head->get_Next();
		}
		$this->num_nodes--;
		return $current->get_Data();
    }
	
    # adds node to end of list, storing given data in node
	function ll_append($data){
		$new_node = new Node($data);
		
		if($this->num_nodes == 0){
			$this->head = $new_node;
			$this->tail = $new_node;
		}
		else{
			$last_node = $this->tail;
			$last_node->set_Next($new_node);
			$this->tail = $new_node;
		}
		$this->num_nodes++;
	}
	
    # removes first node in list THAT MATCHES target
    # returns pointer to the data UNLESS node does not exist (returns nil)
	function ll_remove($target, $compfunc){
		if( $this->num_nodes == 0){
			return NULL;
		}
		
		$current = $this->head;
		$num_curr = 0;
		
		while($num_curr < $this->num_nodes){
			$exists = $compfunc( $target, $current->get_Data() );
			# if finds node with value to remove
			if($exists == 1){
				# if first node
				if($num_curr == 0){
					# if only one node in list
					if($$this->num_nodes == 1){
						$this->head = NULL;
						$this->tail = NULL;
					}
					else{
						$this->head = $this->head->get_Next();
					}
				}
				# if any other node
				else{
					# get node that comes before one deleting
					$prev = $this->head;
					for($x=0; $x<$num_curr-1; $x++){
						$prev = $prev->get_Next();
					}
					# if last node
					if($num_curr == $this->num_nodes-1){
						$prev->set_Next(NULL);
						$this->tail = $prev;
					}
					# if middle node
					else{
						$prev->set_Next($prev->get_Next()->get_Next());
					}
				}
				$this->num_nodes--;
				return $current->get_Data();
			}
			$num_curr++;
			$current = $current->get_Next();
		}
		return NULL;
	}
	 
     # removes all nodes from list
	 function ll_clear(){
		 $this->head = NULL;
		 $this->tail = NULL;
		 $this->num_nodes = 0;
	 }
	 
     # traverses list and applies given function to the data in each node
     function ll_map($function) {
		 $current = $this->head;
		 $num_curr = 0;
		 
		 while($num_curr < $this->num_nodes){
			 $current->set_Data( $function($current->get_Data()) );
			 $current = $current->get_Next();
			 $num_curr++;
		 }
     }
	 
     # returns size of linked list
     function ll_size(){
		 return $this->num_nodes;
     }
	
}


# creates new LinkedList, initializes, and returns
function ll_create(){
	return new Linked_List();
}

function print_values($val){
	echo $val, "\n";
	return $val;
}

# compares value of two different items 
# if same -> returns 1, else returns 0
function compare($target, $node_val){
	 if($target === $node_val){
		 return 1;
	 }
	 return 0;
 }
 
 function add_ten($node_val){
	 return $node_val + 10;
 }


# TESTING

# initialize new Linked List 
$ll = ll_create();

# add different data type values to Linked list
$ll->ll_push(1);
$ll->ll_push(2.89);
$ll->ll_push(3);
$ll->ll_push(4.11);
$ll->ll_append("Sally");
$ll->ll_append("Joe");
$ll->ll_append("Steven");

# print out values of nodes in order - test mapping
echo "A: LinkedList has ", $ll->ll_size(), " nodes\n";
$ll->ll_map(print_values);

# print and delete certain values/nodes
echo "\nRemoving 9 -> ", $ll->ll_remove(9, "compare"), "\n";     # nil (aka blank)
echo "Removing Joe -> ", $ll->ll_remove("Joe", "compare"), "\n";  # Joe (which means able to delete)
echo "Popping node -> ", $ll->ll_pop(), "\n";          # 4.11 
echo "Popping node -> ", $ll->ll_pop(), "\n";          # 3


# print out values of nodes in order - test removing
echo "\nB: LinkedList has ", $ll->ll_size(), " nodes\n";
$ll->ll_map(print_values);


$ll->ll_map(add_ten);
echo "\nC: LinkedList has ", $ll->ll_size(), " nodes\n";
$ll->ll_map(print_values);

$ll->ll_clear();
echo "\nD: LinkedList has ", $ll->ll_size(), " nodes\n";
$ll->ll_map(print_values);


?>